### ISSUE TITLE
Optional `requiredFeatures`: `suggestedFeatures` that only improve remote scheduling

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#8316]
**Is your feature request related to a problem? Please describe.**

Some features are not technically required, but should be scheduled on a machine with the feature for performance or other reasons.

For instance `kvm` speeds up the construction of images with a file system on it, but the build process also works when `kvm` is not available.

This does add entropy to the build environment, which is not great.

**Describe the solution you'd like**

A new attribute `suggestedFeatures` that only improves remote scheduling.

If a machine with the listed features is available, schedule it there.
If such a machine is not available, build it wherever.

**Describe alternatives you've considered**

Remove the `kvm` flag from such builds.
This will likely slow down builds on build farms with mixed features. For instance, it may consist of bare metal machines with `kvm` that handle VM stuff and base load, plus autoscaled VMs to handle regular builds and peak load. VMs generally don't support nested virtualization, so `kvm` builds should not be scheduled onto them.

**Additional context**

- Blocks necessary improvement to https://github.com/NixOS/nixpkgs/pull/229361

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by Rosuavio** [comment]
## Tl;dr
I think this is a great idea

Currently nix-install-action's repo recommends telling nix that the system supports kvm when it does not in order to get nixos tests to work on hosts like GitHub actions.

https://github.com/cachix/install-nix-action/tree/35806937f12c927d9c7223431c7261e2eb2a33ef#how-do-i-run-nixos-tests

I personally had to learn a lot more about GitHub actions, system features and nix configuration, then I wanted to understand why custom nixos tests were not building on GitHub when they did on my machine.

After I figured all of this out, it seemed like this is a known paper cut that everyone has been dealing with (according to my interpretation of the situation)

I think this change could help make the NixOS testing infra inside nixpkgs a lot more usable outside of nixpkgs for a common set of users (people using GitHub actions).

Also I worry that the current solution to this problem (recommending users tell nix that the system supports features it does not support) could lead to other errors when it is actually needed.

**Comment by thufschmitt** [comment]
Discussed during the Nix team meeting. Approved the idea, contributions welcome :)

<details><summary>Discussion log</summary>

- @thufschmitt: Maybe redesign how we schedule things?
- @roberth: What needs to be added?
- @thufschmitt: The implementation of scheduling and remote scheduling should be redone
- @roberth: That seems like an implementation level concern, not an addition to the derivation format
- @edolstra: Do we have more use cases than kvm?
- @thufschmitt: big-parallel
- @edolstra: KVM may be a bad example because with support the build tends to time out
- @ericson2315: Maybe it should be a map where the values says whether is required or suggested
- @thufschmitt: Bike shedding can be done later
- Conclusion: idea approved
</details>
